nibabel/cifti2/cifti2.py

```diff
@@ -0,0 +1,848 @@
+CIFTI_BrainStructures = ('CIFTI_STRUCTURE_ACCUMBENS_LEFT',
+                         'CIFTI_STRUCTURE_ACCUMBENS_RIGHT',
+                         'CIFTI_STRUCTURE_ALL_WHITE_MATTER',
+                         'CIFTI_STRUCTURE_ALL_GREY_MATTER',
+                         'CIFTI_STRUCTURE_AMYGDALA_LEFT',
+                         'CIFTI_STRUCTURE_AMYGDALA_RIGHT',
+                         'CIFTI_STRUCTURE_BRAIN_STEM',
+                         'CIFTI_STRUCTURE_CAUDATE_LEFT',
+                         'CIFTI_STRUCTURE_CAUDATE_RIGHT',
+                         'CIFTI_STRUCTURE_CEREBELLAR_WHITE_MATTER_LEFT',
+                         'CIFTI_STRUCTURE_CEREBELLAR_WHITE_MATTER_RIGHT',
+                         'CIFTI_STRUCTURE_CEREBELLUM',
+                         'CIFTI_STRUCTURE_CEREBELLUM_LEFT',
+                         'CIFTI_STRUCTURE_CEREBELLUM_RIGHT',
+                         'CIFTI_STRUCTURE_CEREBRAL_WHITE_MATTER_LEFT',
+                         'CIFTI_STRUCTURE_CEREBRAL_WHITE_MATTER_RIGHT',
+                         'CIFTI_STRUCTURE_CORTEX',
+                         'CIFTI_STRUCTURE_CORTEX_LEFT',
+                         'CIFTI_STRUCTURE_CORTEX_RIGHT',
+                         'CIFTI_STRUCTURE_DIENCEPHALON_VENTRAL_LEFT',
+                         'CIFTI_STRUCTURE_DIENCEPHALON_VENTRAL_RIGHT',
+                         'CIFTI_STRUCTURE_HIPPOCAMPUS_LEFT',
+                         'CIFTI_STRUCTURE_HIPPOCAMPUS_RIGHT',
+                         'CIFTI_STRUCTURE_OTHER',
+                         'CIFTI_STRUCTURE_OTHER_GREY_MATTER',
+                         'CIFTI_STRUCTURE_OTHER_WHITE_MATTER',
+                         'CIFTI_STRUCTURE_PALLIDUM_LEFT',
+                         'CIFTI_STRUCTURE_PALLIDUM_RIGHT',
+                         'CIFTI_STRUCTURE_PUTAMEN_LEFT',
+                         'CIFTI_STRUCTURE_PUTAMEN_RIGHT',
+                         'CIFTI_STRUCTURE_THALAMUS_LEFT',
+                         'CIFTI_STRUCTURE_THALAMUS_RIGHT')
+
+
+def _value_if_klass(val, klass, none_ok=True):
+    if none_ok and val is None:
+        return val
+    elif isinstance(val, klass):
+        return val
+    else:
+        raise ValueError('Not a valid %s instance.' % klass.__name__)
+
+
+def _value_or_make_klass(val, klass):
+    if val is None:
+        return klass()
+    else:
+        return _value_if_klass(val, klass)
+
+
+def _underscore(string):
+    """ Convert a string from camelcase to underscored """
+    string = re.sub(r'([A-Z]+)([A-Z][a-z])', r'\1_\2', string)
+    return re.sub(r'([a-z0-9])([A-Z])', r'\1_\2', string).lower()
+
+
+class Cifti2MetaData(xml.XmlSerializable):
+    """ A list of key-value pairs stored in the list self.data """
+
+    def __init__(self, nvpair=None):
+        self.data = []
+        self.add_metadata(nvpair)
+
+    def _add_remove_metadata(self, metadata, func):
+        pairs = []
+        if isinstance(metadata, dict):
+            pairs = metadata.items()
+        elif isinstance(metadata, (list, tuple)):
```

Comment:
Do you have a reason to restrict to (list, tuple) here?   Why not any sequence, such as an array?

Maybe something like:

```
    def _check_nvpairs(self, metadata):
        msg = ('nvpair must be a dict-like, a length 2 sequence or '
               'sequence of length 2 sequences')
        if hasattr(metadata, 'items'):
            return self._check_nvpairs(metadata.items())
        metadata = list(metadata)  # Raises error for non-sequence
        if isinstance(metadata[0], string_types):
            if len(metadata) != 2:
                raise ValueError(msg)
            return [tuple(metadata)]
        # Sequence of sequence case
        return sum([self._check_nvpairs(pair) for pair in metadata], [])

    def add_metadata(self, metadata):
        if metadata is None:
            return
        for pair in self._check_nvpairs(metadata):
            if pair not in self.data:
                self.data.append(pair)
```